### PR TITLE
Fix build for wayland

### PR DESCRIPTION
**Pull request by r0l1**
This is a work in progress pull request to get the wayland issue fixed as soon as possible.
This is based on #199 work.

Right now the build progress fails with:

```
# github.com/go-gl/glfw/v3.2/glfw
In file included from ./c_glfw_linbsd.go:16:0:
./glfw/src/wayland-relative-pointer-unstable-v1-client-protocol.c:34:35: error: redefinition of ‘types’
 static const struct wl_interface *types[] = {
                                   ^~~~~
In file included from ./c_glfw_linbsd.go:15:0:
./glfw/src/wayland-pointer-constraints-unstable-v1-client-protocol.c:37:35: note: previous definition of ‘types’ was here
 static const struct wl_interface *types[] = {
                                   ^~~~~
```

Any hints how to fix this?

Thanks!

### CODE REVIEW

**Comment by r0l1** [comment]
Fixed it. I also removed the generate steps and added a generate script. This is only required for developers. I would update the wiki, if this pull request gets merged.

**Comment by r0l1** [comment]
The [wayland developer wiki](https://github.com/go-gl/glfw/wiki/Development) should be updated to following content if this pull request gets merged. 

### Generating wayland source/header files

A helper script located at `scripts/generate-wayland-protocols.sh` should be used to update the wayland source/header files. The [`wayland-scanner`](https://github.com/wayland-project/wayland) utility is required.

```
scripts/generate-wayland-protocols.sh ./v3.2/glfw/glfw/src
```

**Comment by r0l1** [comment]
# Testing
Install the[ xorg-xeyes utility](https://www.archlinux.org/packages/extra/x86_64/xorg-xeyes/) and run it with:

```
xeyes
```

## X11
```
go build
```

Move the mouse over the window and the xeyes should move, because glfw uses X11. 

## Wayland
```
go build -tags wayland
```

Move the mouse over the window and the xeyes should not move, because glfw uses wayland. 

## Code

```go
package main

import (
	"log"
	"runtime"
	"time"

	"github.com/go-gl/gl/v3.2-core/gl"
	"github.com/r0l1/glfw/v3.2/glfw"
)

const (
	winWidth  = 400
	winHeight = 500
	fps       = time.Second / 30
)

func init() {
	runtime.LockOSThread()
}

func main() {
	if err := glfw.Init(); err != nil {
		log.Fatalln(err)
	}
	glfw.WindowHint(glfw.ContextVersionMajor, 3)
	glfw.WindowHint(glfw.ContextVersionMinor, 2)
	glfw.WindowHint(glfw.OpenGLProfile, glfw.OpenGLCoreProfile)
	glfw.WindowHint(glfw.OpenGLForwardCompatible, glfw.True)

	win, err := glfw.CreateWindow(winWidth, winHeight, "Nuklear Demo", nil, nil)
	if err != nil {
		log.Fatalln(err)
	}
	win.MakeContextCurrent()

	width, height := win.GetSize()
	log.Printf("glfw: created window %dx%d", width, height)

	if err := gl.Init(); err != nil {
		log.Fatalln("opengl: init failed:", err)
	}
	gl.Viewport(0, 0, int32(width), int32(height))

	for !win.ShouldClose() {
		win.SwapBuffers()
		glfw.PollEvents()
		time.Sleep(fps)
	}
}
```

**Comment by tapir** [comment]
Works as intended for me. Though I get this error: PlatformError: Wayland: Focusing a window requires user interaction

**Comment by r0l1** [comment]
I have the same problem and after some research I found two possible causes:

1. [glew with EGL required](https://github.com/kovidgoyal/kitty/issues/157)
2. [No window decorations or close callback on Wayland](https://github.com/glfw/glfw/issues/1068)

It seams, that glfw's wayland support has improved a lot in the new upcoming 3.3 release and that we'll have to wait for it's first public release. Nevertheless I suggest to add initial wayland support to this package as initial base for 3.3.

**Comment by tapir** [comment]
I'm OK with merging this one so at least we have 'some' wayland support for 3.2
Any comments @shurcooL @pwaller ?

**Review by dmitshur — APPROVED** [review]
I don't use Linux on desktop day to day, so I can't review this very thoroughly.

However, from a general look, I don't see any major issues (just minor things that aren't worth blocking the PR for). The changes are isolated to Wayland, and it currently doesn't work, so I don't have objections to moving forward with this.

LGTM.

**Comment by dmitshur** [comment]
Thanks @r0l1 and @thomasruiz for working on this, and @tapir for reviewing!

**Comment by r0l1** [comment]
Thanks for merging!  Hope we'll have full wayland support in 3.3...

File: v3.2/glfw/glfw/src/wl_window.c [hunk 1]
```diff
@@ -24,8 +24,6 @@
 //
 //========================================================================
 
-#define _GNU_SOURCE
-
```
Comment (pwaller) [inline v3.2/glfw/glfw/src/wl_window.c:28]: @r0l1 I'm late to this party - is there are reason to diverge from upstream and remove `_GNU_SOURCE` here?

Context is: https://github.com/go-gl/glfw/pull/271.

Comment (pwaller) [reply v3.2/glfw/glfw/src/wl_window.c:28]: FYI I'm reinstating _GNU_SOURCE here, which documents what you should do if you want to remove it again: https://github.com/go-gl/glfw/pull/271/commits/bac7f0e2e1cc1b17dfc90d2c83c04f174e53acf8

**Pull request merged**